csrc/id_model/indexing.cpp

```diff
@@ -97,128 +98,214 @@ Val* getStrideOfGlobalMemoryTensor(TensorView* tv, int64_t alloc_dim) {
-        }
-
-        // If the position is left of the inlining position, no need to
-        // allocate the domain unless it's shared. For example, if this
-        // is a Shared tensor and the domain is parallelized with TID,
-        // even if it's outside of the CA position, since the domain
-        // is shared, it must be allocated.
-        if (i < inlining_pos &&
-            !ir_utils::isMemorySharedAcross(tv->getMemoryType(), pt)) {
-          continue;
-        }
+// TODO: Refactor this and the allocation lowering pass
+class AllocationDomainSetup : private kir::IrVisitor {
+ public:
+  using IrVisitor::dispatch;
+
+  // Set allocation domain info for all tensors
+  void setup(const std::vector<Expr*>& exprs) {
+    // Find out correct allocation domains for all consumer
+    // tensors. Input tensors are handled after this
+    for (auto expr : exprs) {
+      dispatch(expr);
+    }
 
-        allocation_domains.push_back(loop_id);
+    // Make sure all tensors have allocation domains
+    for (TensorView* producer_tv : used_as_producer) {
+      auto it = tv_alloc_info_map.find(producer_tv);
+      if (it != tv_alloc_info_map.end()) {
+        continue;
       }
-      // Assume Local and Shared are always fully contiguous
-      contiguity =
-          std::vector<std::optional<bool>>(allocation_domains.size(), true);
+
+      // Not yet set. This must be an input tensor.
+      NVF_ERROR(
+          producer_tv->isFusionInput(),
+          "Expected a fusion input: ",
+          producer_tv->toString());
+
+      // For fusion input, we can just use getMaybeAllocationDomain.
+
+      auto alloc_info = getIndexingAllocationInfo(
+          producer_tv,
+          producer_tv->getMaybeAllocationDomain(),
+          producer_tv->domain()->contiguity());
+
+      tv_alloc_info_map.emplace(producer_tv, alloc_info);
     }
   }
 
-  std::vector<IterDomain*> promoted_allocation_domains;
-  promoted_allocation_domains.reserve(allocation_domains.size());
-
-  // Loop promotion may affect allocations. Promotions of intermediate
-  // domains may not be defined correctly. Only consider loop domains
-  // for now.
-  for (auto& allocation_domain : allocation_domains) {
-    bool is_loop = std::find(
-                       tv->getLoopDomain().begin(),
-                       tv->getLoopDomain().end(),
-                       allocation_domain) != tv->getLoopDomain().end();
-    IterDomain* promotion_domain = nullptr;
-    if (is_loop) {
-      promotion_domain = getLoopPromotion(allocation_domain, id_model);
+  void dispatch(Expr* expr) override {
+    if (ir_utils::isTvOp(expr)) {
+      for (auto out_tv : ir_utils::filterByType<TensorView>(expr->outputs())) {
+        // Note that since we are dealing with a Kernel IR, a single
+        // tensor may show up as consumers multiple times, e.g.,
+        // zero initialization and actual definition. Using the last
+        // expr should give us correct allocation info.
```

Comment:
Out of curiosity: is there anything special about the last expr? It seems like they would all have the same allocation info.

Reply:
Here's the lowering result from the `InlinedUnroll` test:

```
FOR i70 in iS8{( ceilDiv(3, 1) )}:
  T2_l[ iS3{4} ] ca_pos( 1 ) = ALLOCATE(buffer=T2_l[ iS3{4} ] ca_pos( 1 ), mem_type=register, size=4, zero_init=false, resets_to_zero=false)
  FOR i69 in iS7{4}:
    IF Inline true:
      T2_l[i69] view( T2 )
         = Set( 0, cache_op=Streaming )
  T3_l[ bS12{( ceilDiv(1, 1) )}, bUR13{1}, iS5{4} ] ca_pos( 3 ) produce_pos( 3 ) = ALLOCATE(buffer=T3_l[ bS12{( ceilDiv(1, 1) )}, bUR13{1}, iS5{4} ] ca_pos( 3 ) produce_pos( 3 ), mem_type=register, size=4, zero_init=false, resets_to_zero=false)
  IF Unswitch ( ( i70 >= 0 ) && ( i70 < 3 ) ):
    FOR i71 in iUR9{1}:
      FOR i69 in iS7{4}:
        IF Manual true:
          T2_l[i69] view( T2 )
             = Set( T0_g[i69] view( T0 ), cache_op=Streaming )
        IF Manual true:
          T3_l[i69] view( T3 )
             = broadcast( T2_l[i69] view( T2 ) )
        IF Manual true:
          T4_g[( ( i70 * 4 ) + i69 )] view( T4 )
             = T3_l[i69] view( T3 )
             + T1_g[( ( i70 * 4 ) + i69 )] view( T1 );
  ELSE:
    FOR i71 in iUR9{1}:
      FOR i69 in iS7{4}:
        IF Inline ( ( i69 >= 0 ) && ( i69 < 4 ) ):
          T2_l[i69] view( T2 )
             = Set( T0_g[i69] view( T0 ), cache_op=Streaming )
        IF Inline true:
          T3_l[i69] view( T3 )
             = broadcast( T2_l[i69] view( T2 ) )
        IF Inline ( ( ( ( i70 >= 0 ) && ( i70 < 3 ) ) && ( i69 >= 0 ) ) && ( i69 < 4 ) ):
          T4_g[( ( i70 * 4 ) + i69 )] view( T4 )
             = T3_l[i69] view( T3 )
             + T1_g[( ( i70 * 4 ) + i69 )] view( T1 );
```

As you can see, there are three assignments to `T2`. The first one is zero initialization and that has a different set of for-loops than the other two. If we used the loop-nest at the time of the initialization, we would get different allocation iter domains because there's no unrolled loop there.

Reply:
Added more comments